### PR TITLE
#12256: Migrate moreh_cumsum operation from tt_eager to ttnn

### PR DESCRIPTION
**Pull request by ngohoang34**
### Ticket
Link to Github Issue: [#12256](https://github.com/tenstorrent/tt-metal/issues/12256)

### Problem description
`moreh_cumsum` was deprecated alongside `tt_dnn`. This PR ports it to `ttnn` using its operation format.

### What's changed
- Removed `moreh_cumsum`'s tt_dnn implementations
- Move device code to `ttnn`
- Create new wrapper code for `ttnn` with new modules

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/11247560212
- [x] Blackhole Post commit (if applicable) - NA
- [x] Model regression CI testing passes (if applicable) - NA
- [x] New/Existing tests provide coverage for changes 16/16

### CODE REVIEW

**Comment by razorback3** [comment]
Is this PR working on fixing post-commit tests?

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_program_factory.cpp [hunk 1]
```diff
@@ -65,50 +71,50 @@ operation::ProgramWithCallbacks moreh_cumsum_nc(
...
+    const auto writer_kernel_id = tt::operations::primary::CreateWriteKernel(program, writer_kernel_file, all_cores, writer_compile_time_args);
 
     ////////////////////////////////////////////////////////////////////////////
     //                      ComputeKernel SetUp
     ////////////////////////////////////////////////////////////////////////////
     const std::vector<uint32_t> compute_args_group_1{num_cols_per_core_group_1};
     std::map<string, string> compute_defines;
-    const auto compute_kernel_file = "ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/moreh_cumsum/kernels/moreh_cumsum_nc.cpp";
-    const auto compute_kernel_1_id = CreateComputeKernel(
+    const auto compute_kernel_file = "ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/kernels/reader_moreh_cumsum_nc.cpp";
```
Comment (dongjin-na) [inline ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_program_factory.cpp:110]: I think this is the root cause of jit build error. Would you check it?

[the remark this answers is not in the document]
Comment (ngohoang34) [reply ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_program_factory.cpp]: I've fixed it

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/moreh_cumsum.hpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+};
+
+struct MorehCumsumBackward {
+    static Tensor invoke(const Tensor& input, const Tensor& output, const int64_t dim);
+};
+}  // namespace ttnn::operations::moreh::moreh_cumsum
+
+namespace ttnn {
+constexpr auto moreh_cumsum =
+    ttnn::register_operation<"ttnn::moreh_cumsum", ttnn::operations::moreh::moreh_cumsum::MorehCumsum>();
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/moreh_cumsum.hpp:22]:
```suggestion
    ttnn::register_operation_with_auto_launch_op<"ttnn::moreh_cumsum", ttnn::operations::moreh::moreh_cumsum::MorehCumsum>();
```

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/moreh_cumsum.hpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+constexpr auto moreh_cumsum_backward =
+    ttnn::register_operation<"ttnn::moreh_cumsum_backward", ttnn::operations::moreh::moreh_cumsum::MorehCumsumBackward>();
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/moreh_cumsum.hpp:25]:
```suggestion
    ttnn::register_operation_with_auto_launch_op<"ttnn::moreh_cumsum_backward", ttnn::operations::moreh::moreh_cumsum::MorehCumsumBackward>();
```

**Review by hschoi4448 — APPROVED** [review]

**Review by razorback3 — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+#include "tt_dnn/op_library/moreh_helper_functions.hpp"
+#include "ttnn/operation.hpp"
+#include "ttnn/tensor/tensor.hpp"
+
+namespace ttnn::operations::moreh::moreh_cumsum {
+void MorehCumsumDeviceOperation::validate_inputs(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    const auto dim = operation_attributes.dim;
+    TT_ASSERT((dim >= 0 && dim <= 3), "dim should be 0 - 3");
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp:17]: TT_ASSERT is removed in release builds..
Please use TT_FATAL and include the value of the dim in the message for easier debugging.

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    const auto& output = tensor_args.output;
+
+    auto input_shape = input.get_shape();
+    const auto& output_shape = output.get_shape();
+    auto input_shape_wo_padding = input.get_shape().value.without_padding();
+    const auto& output_shape_wo_padding = output.get_shape().value.without_padding();
+
+    for (int i = 0; i < input_shape.rank(); ++i) {
+        TT_ASSERT(input_shape[i] == output_shape[i]);
+        TT_ASSERT(input_shape_wo_padding[i] == output_shape_wo_padding[i]);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp:28]: Please replase with TT_FATAL and add an explaining error description.
See https://github.com/tenstorrent/tt-metal/blob/main/ErrorMessageBestPractices.md

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    validate_inputs(operation_attributes, tensor_args);
+};
+
+MorehCumsumDeviceOperation::shape_return_value_t MorehCumsumDeviceOperation::compute_output_shapes(
+    const operation_attributes_t&, const tensor_args_t& tensor_args) {
+    const auto& input_tensor = tensor_args.input;
+    auto output_shape = input_tensor.get_shape().value;
+    return ttnn::Shape{tt::tt_metal::Shape(output_shape)};
+    };
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp:52]:
```suggestion
}
```

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+void MorehCumsumDeviceOperation::validate_on_program_cache_hit(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    validate_inputs(operation_attributes, tensor_args);
+};
+
+MorehCumsumDeviceOperation::shape_return_value_t MorehCumsumDeviceOperation::compute_output_shapes(
+    const operation_attributes_t&, const tensor_args_t& tensor_args) {
+    const auto& input_tensor = tensor_args.input;
+    auto output_shape = input_tensor.get_shape().value;
+    return ttnn::Shape{tt::tt_metal::Shape(output_shape)};
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp:51]: can you just do this?
```suggestion
    return input_tensor.get_shape();
```

or 
```suggestion
    auto output_shape = input_tensor.get_shape();
    return output_shape;
```

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/moreh_cumsum.cpp [hunk 8]
```diff
@@ -0,0 +1,14 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "moreh_cumsum.hpp"
+namespace ttnn::operations::moreh::moreh_cumsum {
+Tensor MorehCumsum::invoke(const Tensor& input, const Tensor& output, const int64_t dim) {
+    return ttnn::prim::moreh_cumsum(input, output, dim, false);
```
[the remark this answers is not in the document]
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/moreh_cumsum.cpp:8]: I do not expect changes to it in this PR.

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp (hunk 4 above)
Comment (ngohoang34) [reply ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp]: I've updated the code to use TT_FATAL and include the dimension value in the message for better debugging

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp (hunk 5 above)
Comment (ngohoang34) [reply ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp]: Resolved

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp (hunk 7 above)
Comment (ngohoang34) [reply ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp]: Thanks for pointing that out. I've resolved it

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp (hunk 6 above)
Comment (ngohoang34) [reply ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp]: resolved

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/moreh_cumsum.hpp [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "ttnn/decorators.hpp"
+#include "ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.hpp"
+
+namespace ttnn::operations::moreh::moreh_cumsum {
+struct MorehCumsum {
+    static Tensor invoke(const Tensor& input, const Tensor& output, const int64_t dim);
```
[the remark this answers is not in the document]
Comment (ngohoang34) [reply ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/moreh_cumsum.hpp:12]: Yes, I’ve already checked `TT_FATAL((dim >= 0 && dim <= 3), 'dim should be 0 - 3, but got: {}', dim);` in `moreh_cumsum_device_operation.cpp`

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/moreh_cumsum.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (ngohoang34) [reply ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/moreh_cumsum.cpp:8]: I’ll make a note of this and consider it for future changes

**Comment by ngohoang34** [comment]
@ayerofieiev-tt, since I have resolved your comments, could you please re-review this PR?

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_program_factory.cpp [hunk 10]
```diff
@@ -1,53 +1,55 @@
...
-namespace tt {
-using namespace constants;
-namespace operations {
+namespace ttnn::operations::moreh::moreh_cumsum {
 
-namespace primary {
+MorehCumsumDeviceOperation::ProgramFactory::cached_program_t MorehCumsumDeviceOperation::ProgramFactory::create(
+    const operation_attributes_t& operation_attributes,
+    const tensor_args_t& tensor_args,
+    tensor_return_value_t& output_tensor) {
```
[the remark this answers is not in the document]
Comment (mrshaw01) [reply ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_program_factory.cpp:16]: same for override_runtime_arguments

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_program_factory.cpp [hunk 11]
```diff
@@ -171,31 +173,36 @@ operation::ProgramWithCallbacks moreh_cumsum_nc(
...
+            runtime_args[0] = input_buffer->address();
         }
-    };
 
-    return {.program = std::move(program), .override_runtime_arguments_callback = override_runtime_arguments_callback};
+        {
+            auto& runtime_args = GetRuntimeArgs(program, unary_writer_kernel_id, core);
+            runtime_args[0] = output_buffer->address();
+        }
+    }
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_program_factory.cpp:206]: Suggest change:
```
const auto* input_address = input.buffer()->address();
const auto* output_address = output.buffer()->address();
for (uint32_t i = 0; i < num_cores; ++i) {
    CoreCoord core = {i / num_cores_y, i % num_cores_y};
    {
        auto& runtime_args = GetRuntimeArgs(program, unary_reader_kernel_id, core);
        runtime_args[0] = input_address;
    }

    {
        auto& runtime_args = GetRuntimeArgs(program, unary_writer_kernel_id, core);
        runtime_args[0] = output_address;
    }
}
```

**Review by mrshaw01 — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_moreh_cumsum.py [hunk 12]
```diff
@@ -74,11 +79,7 @@ def test_moreh_cumsum_dim(input_shape, dim, device):
 
     cpu_layout = ttnn.ROW_MAJOR_LAYOUT
     tt_output_cpu = (
-        ttnn.experimental.operations.primary.moreh_cumsum(tt_input, tt_output, dim=dim)
-        .cpu()
-        .to(cpu_layout)
-        .unpad_from_tile(output_shape)
-        .to_torch()
+        ttnn.moreh_cumsum(tt_input, tt_output, dim=dim).cpu().to(cpu_layout).unpad_from_tile(output_shape).to_torch()
```
[the remark this answers is not in the document]
Comment (ngohoang34) [reply tests/ttnn/unit_tests/operations/test_moreh_cumsum.py]: resolved

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+};
+
+void MorehCumsumDeviceOperation::validate_on_program_cache_hit(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    validate_inputs(operation_attributes, tensor_args);
+};
+
+MorehCumsumDeviceOperation::shape_return_value_t MorehCumsumDeviceOperation::compute_output_shapes(
+    const operation_attributes_t&, const tensor_args_t& tensor_args) {
+    const auto& input_tensor = tensor_args.input;
```
[the remark this answers is not in the document]
Comment (ngohoang34) [reply ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.cpp]: resolved

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.hpp [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+            std::size_t num_cores;
+            std::size_t num_cores_y;
+        };
+
+        using cached_program_t = ttnn::device_operation::CachedProgram<shared_variables_t>;
+
+        static cached_program_t create(
+            const operation_attributes_t& operation_attributes,
+            const tensor_args_t& tensor_args,
+            tensor_return_value_t& output_tensor);
```
[the remark this answers is not in the document]
Comment (ngohoang34) [reply ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.hpp]: resolved

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.hpp [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+
+        static void override_runtime_arguments(
+            cached_program_t& cached_program,
+            const operation_attributes_t& operation_attributes,
+            const tensor_args_t& tensor_args,
+            tensor_return_value_t& output_tensor);
```
[the remark this answers is not in the document]
Comment (ngohoang34) [reply ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.hpp]: resolved

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_program_factory.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (ngohoang34) [reply ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_program_factory.cpp]: resolved

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_program_factory.cpp (hunk 11 above)
Comment (ngohoang34) [reply ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_program_factory.cpp]: resolved

File: ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/moreh_cumsum.hpp [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+#pragma once
+
+#include "ttnn/decorators.hpp"
+#include "ttnn/operations/moreh/moreh_cumsum/device/moreh_cumsum_device_operation.hpp"
+
+namespace ttnn::operations::moreh::moreh_cumsum {
+struct MorehCumsum {
+    static Tensor invoke(
+        const Tensor& input,
+        const Tensor& output,
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/moreh_cumsum.hpp:14]: I am confused. Is this a preallocated output? 
If so
* why is it const& here and why is it not optional here?
* why do you always create an output tensor here https://github.com/tenstorrent/tt-metal/pull/12296/files#diff-b4df615d6d6163c2ed02bef119ca5b210010279596c6f7841c41df470f29e0c8R63 ?

Comment (ngohoang34) [reply ttnn/cpp/ttnn/operations/moreh/moreh_cumsum/moreh_cumsum.hpp]: I've modified the code to address your comment

**Review by ayerofieiev-tt — APPROVED** [review]

**Pull request merged**